### ISSUE TITLE
Strange behaviour of REPL when running first few Ch3 examples

### ISSUE THREAD

**Issue by ljalbinson** [databricks/Spark-The-Definitive-Guide#28]
If I run the following code from Chapter 3 as two separately pasted snippets all is OK.

`// in Scala - snippet 1
import spark.implicits._
case class Flight(DEST_COUNTRY_NAME: String,
                  ORIGIN_COUNTRY_NAME: String,
                  count: BigInt)
val flightsDF = spark.read
  .parquet("/data/flight-data/parquet/2010-summary.parquet/")
val flights = flightsDF.as[Flight]

// COMMAND ----------

// in Scala - snippet 2
flights
  .filter(flight_row => flight_row.ORIGIN_COUNTRY_NAME != "Canada")
  .map(flight_row => flight_row)
  .take(5)`

If I run it as one pasted snippet the result is a java exception.

`java.lang.NullPointerException
  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)`

I am using Spark 2.2.2.

Kind regards, Lawrence

**Comment by anabranch** [comment]
Not sure what the issue is, since it still works (when they're separate) I'm just going to close this out.

**Comment by ljalbinson** [comment]
I appreciate it is not an issue for your book.
However, isn't it a little worrying that the combined case fails? I don't think it should. But I'm happy to be corrected.
And, of course, I'm happy to report it elsewhere if that is more appropriate (assuming it is a bug and not my lack of understanding).
Kind regards, Lawrence

**Comment by anabranch** [comment]
It's probably something to do with the implicits and the REPL compiler, could you try running the first line alone (import ...) and then running the rest as one block?

**Comment by ljalbinson** [comment]
Putting the import first doesn't solve the problem; however, it leads to a useful error message, namely:

`<console>:18: error: Unable to find encoder for type stored in a Dataset.  Primitive types (Int, ....`

In [this](https://stackoverflow.com/questions/38664972/why-is-unable-to-find-encoder-for-type-stored-in-a-dataset-when-creating-a-dat) post the case class is moved out of the current scope and the problem is solved, so it seems that it is an implicit resolution issue.

I have tried this with the Ch3 example and it works.

So, in summary, declare the case class as a first snippet and then declare the import, the two vals and the filter/map/take as a second snippet and it is fine.

**Comment by anabranch** [comment]
There we are. Glad we figured it out!
